### PR TITLE
Upgrade to hassil 2.0

### PR DESCRIPTION
**Pull request by synesthesiam**
Bumps hassil to 2.0.0: https://github.com/home-assistant/hassil/compare/1.7.4...v2.0.0

There are some significant changes in hassil 2.0, fixing several outstanding bugs. I've done my best to fix the existing sentences, but I need help from some language leaders.

## _test_failures.yaml

I've added a new file to `tests/<language>` called `_test_failures.yaml` which contains a list of sentences that are being tested for but are failing for some reason. Language leaders should start here when looking at their language for this PR. Once fixed, please remove the sentence from `_test_failures.yaml`. 

## Matching Changes

For testing, the biggest change is going from the **first** sentence matched to the **best** sentence matched. This mirrors the code in HA (and will soon replace it) so that certain matches are prioritized:
- Matches with a `name` slot are highest priority, with the longest matching `name` winning
- Matches with more literal text (the text outside of lists and wildcards) are higher priority
- Matches with wildcards are lower priority

In a few cases, this will cause a different sentence template to be recognized than before. For example, if you have a test sentence "open the garage door" and an entity named "garage door", it will always match the template open the {name} instead of open the {device_class} now.

## Hassil Changes

Some high level hassil changes:
- Casing is now preserved, so list values/wildcards/etc. will no longer be lower-cased
- Sentence templates are filtered with a fuzzy regex before matching is attempted. This speeds up most languages, but slows down a few (those languages have `filter_with_regex: false` in their `_common.yaml` files.
- Extra whitespace around wildcards is removed
- Number words will now be used for all genders, cases, etc. in the language (see [unicode-rbnf release](https://github.com/rhasspy/unicode-rbnf/releases/tag/v2.0.0))

Details of hassil changes:

- Allow wildcards to be followed by expansion rules and lists
- Use regular expressions to filter sentence templates
- Add `filter_with_regex` to intent settings and intent data (`false` disables regex filtering)
- Filter text slot list values by required/excluded context during matching
- Use a trie to filter range slot list values based on remaining text to be matched
- Add `required_keywords` section to intent data to skip sentences without specific keywords
- Preserve case during matching 
- Strip punctuation before text processing
- Remove extraneous whitespace from the end of wildcards
- Refactor string matching code into `string_matcher.py`

### CODE REVIEW

**Review by TheFes — APPROVED** [review]
No further comments about NL changes

File: sentences/en/homeassistant_HassTimerStatus.yaml [hunk 1]
```diff
@@ -8,27 +8,46 @@ intents:
...
       # Status of...
       - sentences:
           - "status of [the|my] timer[s]"
           - "status of [the|my] <timer_start> timer"
           - "status of [the|my] {area} timer[s]"
           - "status of [the|my] timer[s] in <area>"
+        required_keywords:
+          - "timer"
+          - "timers"
+          - "timers"
```
Comment (balloob) [inline sentences/en/homeassistant_HassTimerStatus.yaml:44]: Duplicate?

**Review by balloob — APPROVED** [review]

**Review by andrejs2 — APPROVED** [review]

**Review by LubosKadasi — APPROVED** [review]

**Review by cibernox — APPROVED** [review]

**Review by dontinelli — APPROVED** [review]
Approved from de-CH point of view

**Review by nagyrobi — APPROVED** [review]

**Review by skynetua — APPROVED** [review]
Uk is ok

**Review by tetele — APPROVED** [review]
RO looks good

**Review by davefx — APPROVED** [review]

**Review by Arkkimaagi — APPROVED** [review]
Finnish at a glance looks good, but when I tried to run test for sampling sentences in Finnish, I got errors.

As I'm a bit rusty here, I thought that this might have been broken somehow, but then I checked the current main branch and it had the same problem with Finnish, oh and with English too. So I think this PR does not break that, it's just like that (at least in Codespaces) or I do not know what I'm doing in relation to this.
```
@Arkkimaagi ➜ /workspaces/intents (synesthesiam-20241103-faster-recognizer) $ python3 -m script.intentfest sample --language fi -n 1
{"intent": "HassTurnOff", "text": "kytke päältä valo Keittiö"}
{"intent": "HassGetState", "text": "Paljonko Makuuhuoneen valo alueessa Keittiö on"}
{"intent": "HassMediaPrevious", "text": "edellinen raita laitteelle Makuuhuoneen valo"}
{"intent": "HassMediaUnpause", "text": "jatka laitteen Makuuhuoneen valo toistoa"}
No language set, so cannot convert position digits to words
{"intent": "HassSetPosition", "text": "kytke Makuuhuoneen valo asentoon 0 %"}
No language set, so cannot convert volume_level digits to words
{"intent": "HassSetVolume", "text": "kytke laitteen Makuuhuoneen valo äänenvoimakkuudeksi 0 %"}
{"intent": "HassUnpauseTimer", "text": "jatka minun ajastinta"}
No language set, so cannot convert temperature digits to words
{"intent": "HassClimateSetTemperature", "text": "kytke lämpötila alueessa Keittiö 0 ° celsius"}
{"intent": "HassPauseTimer", "text": "keskeytä minun ajastin"}
{"intent": "HassGetCurrentDate", "text": "kerro mikä mikä päivä nyt on"}
{"intent": "HassTurnOn", "text": "käynnistä Makuuhuoneen valo"}
{"intent": "HassMediaNext", "text": "seuraava raita laitteelle Makuuhuoneen valo"}
{"intent": "HassGetCurrentTime", "text": "kerro mitä kello on juuri nyt"}
No language set, so cannot convert seconds digits to words
No language set, so cannot convert minutes digits to words
No language set, so cannot convert seconds digits to words
No language set, so cannot convert minutes digits to words
No language set, so cannot convert hours digits to words
No language set, so cannot convert minutes digits to words
No language set, so cannot convert seconds digits to words
Terminated
```

So, to sum it up, Finnish seems to be working on this branch at least just as fine as in main branch. Approved on my side.

**Review by piitaya — APPROVED** [review]

**Review by abmantis — APPROVED** [review]

**Comment by tetele** [comment]
https://github.com/home-assistant/intents/pull/2507/commits/0f01d51e96a567a651f55cbdb435a80ede11c7d4 could be its own PR, i guess

**Comment by synesthesiam** [comment]
> [0f01d51](https://github.com/home-assistant/intents/commit/0f01d51e96a567a651f55cbdb435a80ede11c7d4) could be its own PR, i guess

Yes, but I'm ready to get this merged and for everyone to start working independently on it :smile:

File: tests/nb/homeassistant_HassPauseTimer.yaml [hunk 2]
```diff
@@ -25,7 +25,7 @@ tests:
       name: HassPauseTimer
       slots:
         name:
-          - "pizza "
+          - "pizza"
```
Comment (LaStrada) [inline tests/nb/homeassistant_HassPauseTimer.yaml:28]: This line can be removed

**Comment by LaStrada** [comment]
> I've added a new file to tests/<language> called _test_failures.yaml which contains a list of sentences that are being tested for but are failing for some reason. Language leaders should start here when looking at their language for this PR. Once fixed, please remove the sentence from _test_failures.yaml.

Should this file always exist? If I fix the failing test for my language, should I just remove the sentence or the file?

**Review by abmantis — APPROVED** [review]

**Review by LaStrada — APPROVED** [review]
Had a small comment, but that can be fixed in a separate PR

**Review by luyzfernando08 — APPROVED** [review]

**Review by balloob — APPROVED** [review]

**Comment by TheFes** [comment]
> > I've added a new file to tests/ called _test_failures.yaml which contains a list of sentences that are being tested for but are failing for some reason. Language leaders should start here when looking at their language for this PR. Once fixed, please remove the sentence from _test_failures.yaml.
> 
> Should this file always exist? If I fix the failing test for my language, should I just remove the sentence or the file?

As there is no such file for NL, I assume it is not needed if there is nothing to fix

**Comment by andrejs2** [comment]
> ## _test_failures.yaml
> 
> I've added a new file to `tests/<language>` called `_test_failures.yaml` which contains a list of sentences that are being tested for but are failing for some reason. Language leaders should start here when looking at their language for this PR. Once fixed, please remove the sentence from `_test_failures.yaml`.

@synesthesiam I've fixed sentences and tests according to `_test_failures.yaml` file, but in some cases test via Terminal - > Run task... was OK, but if I tested via `python3 -m script.intentfest parse --language sl --sentence 'vključi ventilatorje v kuhinji'` I got an error and vice versa...`sentence expected to fail but it didn't`... 
But now I think I fixed all of the listed sentences and some others too :)

If I remove sentences I get error when testing via Terminal Run task... or should I just remove the file `_test_failures.yaml`?
